In [1]:
import pandas as pd
import numpy as np

from catboost import CatBoostRegressor
cat = CatBoostRegressor()

from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

In [2]:
train = pd.read_csv('G:/내 드라이브/Github/BigContest_Flood-ZERO/고경수/train.csv')
test = pd.read_csv('G:/내 드라이브/Github/BigContest_Flood-ZERO/고경수/test.csv')
submission = pd.read_csv('G:/내 드라이브/Github/BigContest_Flood-ZERO/고경수/submission.csv')

In [3]:
유역평균강수 = ['유역평균강수1','유역평균강수2','유역평균강수3','유역평균강수4','유역평균강수5','유역평균강수6']
강우A = ['강우A1','강우A2','강우A3','강우A4','강우A5','강우A6']
강우B = ['강우B1','강우B2','강우B3','강우B4','강우B5','강우B6']
강우C = ['강우C1','강우C2','강우C3','강우C4','강우C5','강우C6']
강우D = ['강우D1','강우D2','강우D3','강우D4','강우D5','강우D6']
수위E = ['수위E1','수위E2','수위E3','수위E4','수위E5','수위E6']
수위D = ['수위D1','수위D2','수위D3','수위D4','수위D5','수위D6']

In [4]:
train['홍수사상번호'] = train['홍수사상번호'].astype(str)
train['월'] = train['월'].astype(str)
train['일'] = train['일'].astype(str)
train['시간'] = train['시간'].astype(str)

test['홍수사상번호'] = test['홍수사상번호'].astype(str)
test['월'] = test['월'].astype(str)
test['일'] = test['일'].astype(str)
test['시간'] = test['시간'].astype(str)

In [5]:
X_train, X_test, y_train, y_test = train_test_split(train[['홍수사상번호','연','월','일','시간',*유역평균강수, *강우A, *강우B, *강우C, *강우D, *수위E, *수위D]],
                                                           train['유입량'], test_size=0.0681425, shuffle=False)

Original_train = X_train # 사상번호 1 ~ 22
Original_test = X_test # 사상번호 23 ~ 25

---

# 집단 1

In [6]:
X_train = Original_train[[유역평균강수[0], 강우A[0], 강우B[0], 강우C[0], 강우D[0], 수위E[0], 수위D[0]]]
X_test = Original_test[[유역평균강수[0], 강우A[0], 강우B[0], 강우C[0], 강우D[0], 수위E[0], 수위D[0]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2153.2331580	total: 150ms	remaining: 2m 29s
1:	learn: 2076.3798116	total: 153ms	remaining: 1m 16s
2:	learn: 2001.4247041	total: 159ms	remaining: 52.9s
3:	learn: 1932.7873324	total: 164ms	remaining: 40.8s
4:	learn: 1863.1954002	total: 168ms	remaining: 33.4s
5:	learn: 1798.1971417	total: 172ms	remaining: 28.4s
6:	learn: 1734.3797453	total: 177ms	remaining: 25.1s
7:	learn: 1674.5060711	total: 181ms	remaining: 22.4s
8:	learn: 1617.7301287	total: 185ms	remaining: 20.3s
9:	learn: 1563.8420293	total: 188ms	remaining: 18.6s
10:	learn: 1513.6169583	total: 191ms	remaining: 17.2s
11:	learn: 1467.2346414	total: 195ms	remaining: 16s
12:	learn: 1419.2806786	total: 201ms	remaining: 15.2s
13:	learn: 1373.8540798	total: 207ms	remaining: 14.6s
14:	learn: 1332.1592344	total: 218ms	remaining: 14.3s
15:	learn: 1290.6286318	total: 224ms	remaining: 13.8s
16:	learn: 1251.5051718	total: 228ms	remaining: 13.2s
17:	learn: 1215.5519151	total: 232ms	remaining: 12.7s
18:	learn

In [7]:
pred = model.predict(X_test)

In [8]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 774

874.4158551368695

---
# 집단 2

In [9]:
X_train = Original_train[[유역평균강수[1], 강우A[1], 강우B[1], 강우C[1], 강우D[1], 수위E[1], 수위D[1]]]
X_test = Original_test[[유역평균강수[1], 강우A[1], 강우B[1], 강우C[1], 강우D[1], 수위E[1], 수위D[1]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2154.1967893	total: 14.3ms	remaining: 14.3s
1:	learn: 2076.5085512	total: 19.3ms	remaining: 9.62s
2:	learn: 2003.2402524	total: 24.9ms	remaining: 8.29s
3:	learn: 1932.8241489	total: 27.6ms	remaining: 6.88s
4:	learn: 1863.2024942	total: 30.2ms	remaining: 6.02s
5:	learn: 1796.9880982	total: 33.9ms	remaining: 5.61s
6:	learn: 1734.2559610	total: 36.5ms	remaining: 5.18s
7:	learn: 1674.5542938	total: 39.1ms	remaining: 4.84s
8:	learn: 1617.0648288	total: 41.6ms	remaining: 4.58s
9:	learn: 1561.7583433	total: 44.3ms	remaining: 4.39s
10:	learn: 1511.5302157	total: 47.1ms	remaining: 4.24s
11:	learn: 1465.7402840	total: 49.7ms	remaining: 4.09s
12:	learn: 1418.1060640	total: 52.3ms	remaining: 3.97s
13:	learn: 1372.9051850	total: 55ms	remaining: 3.88s
14:	learn: 1330.8789458	total: 58.4ms	remaining: 3.83s
15:	learn: 1290.9105630	total: 61ms	remaining: 3.75s
16:	learn: 1251.6214651	total: 64.9ms	remaining: 3.75s
17:	learn: 1213.4625329	total: 69.1ms	remaining: 3

In [10]:
pred = model.predict(X_test)

In [11]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 912

1050.2883773454657

---
# 집단 3

In [12]:
X_train = Original_train[[유역평균강수[2], 강우A[2], 강우B[2], 강우C[2], 강우D[2], 수위E[2], 수위D[2]]]
X_test = Original_test[[유역평균강수[2], 강우A[2], 강우B[2], 강우C[2], 강우D[2], 수위E[2], 수위D[2]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2153.6418334	total: 6.26ms	remaining: 6.25s
1:	learn: 2075.6440251	total: 13.7ms	remaining: 6.84s
2:	learn: 2002.0750886	total: 26.1ms	remaining: 8.68s
3:	learn: 1934.0167875	total: 29.3ms	remaining: 7.29s
4:	learn: 1864.0437272	total: 38ms	remaining: 7.56s
5:	learn: 1798.0971886	total: 41.8ms	remaining: 6.93s
6:	learn: 1734.9827314	total: 44.8ms	remaining: 6.35s
7:	learn: 1675.0708732	total: 52.8ms	remaining: 6.54s
8:	learn: 1621.5638179	total: 55.6ms	remaining: 6.12s
9:	learn: 1566.5166236	total: 60ms	remaining: 5.94s
10:	learn: 1515.9620365	total: 62.7ms	remaining: 5.63s
11:	learn: 1466.4810021	total: 65.2ms	remaining: 5.37s
12:	learn: 1419.0539019	total: 67.8ms	remaining: 5.15s
13:	learn: 1373.7260941	total: 70.9ms	remaining: 4.99s
14:	learn: 1331.2354660	total: 73.5ms	remaining: 4.82s
15:	learn: 1290.9723157	total: 76.1ms	remaining: 4.68s
16:	learn: 1250.5546915	total: 79.8ms	remaining: 4.62s
17:	learn: 1212.2001233	total: 82.5ms	remaining: 4

In [13]:
pred = model.predict(X_test)

In [14]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 985

1103.7250447786294

---
# 집단 4

In [15]:
X_train = Original_train[[유역평균강수[3], 강우A[3], 강우B[3], 강우C[3], 강우D[3], 수위E[3], 수위D[3]]]
X_test = Original_test[[유역평균강수[3], 강우A[3], 강우B[3], 강우C[3], 강우D[3], 수위E[3], 수위D[3]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2152.4999031	total: 6.26ms	remaining: 6.25s
1:	learn: 2077.9277845	total: 8.84ms	remaining: 4.41s
2:	learn: 2002.7285430	total: 11.4ms	remaining: 3.79s
3:	learn: 1932.9301099	total: 15.1ms	remaining: 3.77s
4:	learn: 1862.4119459	total: 18.2ms	remaining: 3.63s
5:	learn: 1796.8286200	total: 20.9ms	remaining: 3.45s
6:	learn: 1732.5042378	total: 24ms	remaining: 3.41s
7:	learn: 1672.3836280	total: 27ms	remaining: 3.35s
8:	learn: 1614.7215936	total: 30.1ms	remaining: 3.32s
9:	learn: 1560.9591339	total: 33.4ms	remaining: 3.31s
10:	learn: 1510.4969103	total: 39.2ms	remaining: 3.52s
11:	learn: 1463.0953528	total: 41.9ms	remaining: 3.45s
12:	learn: 1414.6526358	total: 44.5ms	remaining: 3.38s
13:	learn: 1370.6487415	total: 48.2ms	remaining: 3.4s
14:	learn: 1327.7705647	total: 52.6ms	remaining: 3.45s
15:	learn: 1286.7067073	total: 55.2ms	remaining: 3.4s
16:	learn: 1246.7797463	total: 59.3ms	remaining: 3.43s
17:	learn: 1208.4902794	total: 64.3ms	remaining: 3.5

In [16]:
pred = model.predict(X_test)

In [17]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 769

817.746221601441

---
# 집단 5

In [18]:
X_train = Original_train[[유역평균강수[4], 강우A[4], 강우B[4], 강우C[4], 강우D[4], 수위E[4], 수위D[4]]]
X_test = Original_test[[유역평균강수[4], 강우A[4], 강우B[4], 강우C[4], 강우D[4], 수위E[4], 수위D[4]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2152.5591258	total: 3.62ms	remaining: 3.62s
1:	learn: 2077.9908038	total: 9.89ms	remaining: 4.93s
2:	learn: 2002.7930556	total: 12.5ms	remaining: 4.16s
3:	learn: 1932.9983819	total: 15.1ms	remaining: 3.76s
4:	learn: 1862.5457074	total: 17.7ms	remaining: 3.52s
5:	learn: 1796.9214438	total: 22.9ms	remaining: 3.79s
6:	learn: 1732.5345726	total: 26.2ms	remaining: 3.72s
7:	learn: 1672.4280072	total: 28.8ms	remaining: 3.57s
8:	learn: 1614.7811705	total: 31.9ms	remaining: 3.51s
9:	learn: 1561.0400562	total: 37.4ms	remaining: 3.7s
10:	learn: 1510.6691537	total: 39.9ms	remaining: 3.59s
11:	learn: 1462.1204953	total: 42.6ms	remaining: 3.5s
12:	learn: 1413.7845952	total: 47.9ms	remaining: 3.63s
13:	learn: 1367.9704921	total: 51ms	remaining: 3.59s
14:	learn: 1323.0325985	total: 53.6ms	remaining: 3.52s
15:	learn: 1280.5171735	total: 56.6ms	remaining: 3.48s
16:	learn: 1241.0431476	total: 62.1ms	remaining: 3.59s
17:	learn: 1202.8032931	total: 64.8ms	remaining: 3

In [19]:
pred = model.predict(X_test)

In [20]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 789

794.8306987606229

---
# 집단 6

In [21]:
X_train = Original_train[[유역평균강수[5], 강우A[5], 강우B[5], 강우C[5], 강우D[5], 수위E[5], 수위D[5]]]
X_test = Original_test[[유역평균강수[5], 강우A[5], 강우B[5], 강우C[5], 강우D[5], 수위E[5], 수위D[5]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2152.4688725	total: 5.56ms	remaining: 5.56s
1:	learn: 2073.9240710	total: 8.79ms	remaining: 4.38s
2:	learn: 1999.9616347	total: 12.8ms	remaining: 4.24s
3:	learn: 1929.8919182	total: 16.2ms	remaining: 4.04s
4:	learn: 1859.9749025	total: 19.7ms	remaining: 3.92s
5:	learn: 1793.1477494	total: 25ms	remaining: 4.15s
6:	learn: 1730.0089208	total: 29.5ms	remaining: 4.18s
7:	learn: 1670.1465408	total: 45.5ms	remaining: 5.64s
8:	learn: 1612.8853789	total: 54.8ms	remaining: 6.04s
9:	learn: 1558.3760149	total: 68.8ms	remaining: 6.81s
10:	learn: 1507.6148127	total: 74.3ms	remaining: 6.68s
11:	learn: 1458.6526207	total: 89.3ms	remaining: 7.35s
12:	learn: 1410.4356653	total: 98.5ms	remaining: 7.48s
13:	learn: 1367.3076846	total: 103ms	remaining: 7.23s
14:	learn: 1324.0772486	total: 126ms	remaining: 8.27s
15:	learn: 1282.0517508	total: 129ms	remaining: 7.94s
16:	learn: 1241.3846938	total: 133ms	remaining: 7.67s
17:	learn: 1204.0828992	total: 138ms	remaining: 7.55

In [22]:
pred = model.predict(X_test)

In [23]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 989

1125.8821272893433

---
# 전체집단

In [24]:
X_train = Original_train[[*유역평균강수, *강우A, *강우B, *강우C, *강우D, *수위E, *수위D]]
X_test = Original_test[[*유역평균강수, *강우A, *강우B, *강우C, *강우D, *수위E, *수위D]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2153.9943977	total: 11.8ms	remaining: 11.8s
1:	learn: 2076.7140606	total: 25.2ms	remaining: 12.6s
2:	learn: 2001.4374847	total: 34.7ms	remaining: 11.5s
3:	learn: 1930.8781025	total: 45.7ms	remaining: 11.4s
4:	learn: 1863.0817933	total: 57.8ms	remaining: 11.5s
5:	learn: 1797.3601656	total: 72.8ms	remaining: 12.1s
6:	learn: 1732.9622734	total: 82.9ms	remaining: 11.8s
7:	learn: 1671.2677140	total: 93.9ms	remaining: 11.6s
8:	learn: 1615.0699156	total: 107ms	remaining: 11.8s
9:	learn: 1560.6271750	total: 120ms	remaining: 11.9s
10:	learn: 1508.1257363	total: 133ms	remaining: 11.9s
11:	learn: 1456.2865439	total: 149ms	remaining: 12.3s
12:	learn: 1411.4579733	total: 159ms	remaining: 12.1s
13:	learn: 1364.5496690	total: 169ms	remaining: 11.9s
14:	learn: 1320.1459654	total: 179ms	remaining: 11.8s
15:	learn: 1278.9890532	total: 191ms	remaining: 11.7s
16:	learn: 1237.6619732	total: 200ms	remaining: 11.6s
17:	learn: 1199.9910644	total: 212ms	remaining: 11.6s
1

In [25]:
pred = model.predict(X_test)

In [26]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 840

1001.5166531826344

---
# 상관계수 가장 높은 것 혼합

In [27]:
train.corr()['유입량']

연         -0.349836
유입량        1.000000
유역평균강수1    0.742268
강우A1       0.617424
강우B1       0.739951
강우C1       0.605588
강우D1       0.700568
수위E1       0.903840
수위D1       0.109475
유역평균강수2    0.752184
강우A2       0.615217
강우B2       0.743754
강우C2       0.626355
강우D2       0.700568
수위E2       0.903840
수위D2       0.124289
유역평균강수3    0.752184
강우A3       0.615217
강우B3       0.771063
강우C3       0.641773
강우D3       0.710927
수위E3       0.903840
수위D3       0.133473
유역평균강수4    0.760476
강우A4       0.612004
강우B4       0.771063
강우C4       0.619889
강우D4       0.722170
수위E4       0.903840
수위D4       0.084623
유역평균강수5    0.767842
강우A5       0.612004
강우B5       0.771063
강우C5       0.619889
강우D5       0.722170
수위E5       0.903840
수위D5       0.091073
유역평균강수6    0.790920
강우A6       0.612004
강우B6       0.767258
강우C6       0.669108
강우D6       0.727581
수위E6       0.903840
수위D6       0.097066
Name: 유입량, dtype: float64

In [28]:
X_train = Original_train[[유역평균강수[5], 강우A[0], 강우B[2], 강우C[5], 강우D[5], 수위E[5], 수위D[2]]]
X_test = Original_test[[유역평균강수[5], 강우A[0], 강우B[2], 강우C[5], 강우D[5], 수위E[5], 수위D[2]]]

categorical_features_indices1 = np.where(X_train.dtypes == np.object)[0]

model = cat
model.fit(X_train, y_train, cat_features = categorical_features_indices1)

Learning rate set to 0.04656
0:	learn: 2151.7329173	total: 3.74ms	remaining: 3.74s
1:	learn: 2073.5252518	total: 6.45ms	remaining: 3.22s
2:	learn: 1999.5838479	total: 9.19ms	remaining: 3.05s
3:	learn: 1929.4329509	total: 11.8ms	remaining: 2.94s
4:	learn: 1859.5434514	total: 14.9ms	remaining: 2.96s
5:	learn: 1792.7494084	total: 17.5ms	remaining: 2.9s
6:	learn: 1728.7837999	total: 20.4ms	remaining: 2.89s
7:	learn: 1669.0336550	total: 23ms	remaining: 2.85s
8:	learn: 1613.7486947	total: 25.6ms	remaining: 2.81s
9:	learn: 1559.8046147	total: 29.4ms	remaining: 2.91s
10:	learn: 1507.9977911	total: 32ms	remaining: 2.88s
11:	learn: 1458.4559812	total: 34.7ms	remaining: 2.85s
12:	learn: 1410.0643152	total: 37.3ms	remaining: 2.83s
13:	learn: 1367.1599186	total: 40.7ms	remaining: 2.86s
14:	learn: 1324.0139572	total: 43.6ms	remaining: 2.86s
15:	learn: 1281.8896640	total: 46.2ms	remaining: 2.84s
16:	learn: 1243.0145127	total: 48.9ms	remaining: 2.83s
17:	learn: 1204.4790101	total: 51.6ms	remaining: 2.

In [29]:
pred = model.predict(X_test)

In [30]:
rmse = (np.sqrt(np.mean(mean_squared_error(y_test, pred))))
rmse
# 1: 910

1034.0869658761287

---

test를 홍수사상 23, 24, 25로 했을 때는 집단 4가 제일 잘나오고

test를 홍수사상 24, 25로 했을 때는 집단 5가 제일 잘나온다..

---

- 홍수사상 23 24 25

1: 774
2: 912
3: 985
4: 769 
5: 789
6: 989 
전체 : 840 
혼합 : 910
    
4 - 1 - 5 - 전체 - 혼합 - 2 - 3 - 6

- 홍수사상 24 25

1: 874 2: 1050 3: 1103 4: 817 5: 794 6: 1125 전체 : 1001 혼합 : 1034

5 - 4 - 1 - 전체 - 혼합 - 2 - 3 - 6